### PR TITLE
Add `StaticTextSlice` kind to `FormatElement` enum

### PR DESCRIPTION
**Pull request by charliermarsh**
Given our current parser abstractions, we need the ability to tell `ruff_formatter` to print a pre-defined slice from a fixed string of source code, which we've introduced here as `FormatElement::StaticTextSlice`.

### CODE REVIEW

**Comment by charliermarsh** [comment]
\cc @MichaReiser

File: crates/ruff_formatter/src/format_element.rs [hunk 1]
```diff
@@ -47,6 +47,13 @@ pub enum FormatElement {
         slice: SyntaxTokenText,
     },
 
+    /// Token constructed by slicing a defined range from a static string
+    StaticTextSlice {
+        text: Rc<str>,
+        start: usize,
+        end: usize,
```
Comment (not-my-profile) [inline crates/ruff_formatter/src/format_element.rs:54]: Shouldn't this rather be `std::ops::Range<usize>`?

Comment (MichaReiser) [inline crates/ruff_formatter/src/format_element.rs:54]: I recommend you to use `TextRange` here. Documents larger than 4GB should be rare. Using `TextRange` helps keeping the overall size of `FormatElement` low which is essential for good formatter performance (fewer bytes to write when formatting, fewer bytes to read when printing, better cache locality...) and memory usage low.

**Comment by charliermarsh** [comment]
I have a builder implemented in #2883 but it hinges on some details from the RustPython AST. Added a generic builder here.

**Pull request merged**